File: script/timing.js
define('timing', ['audio', 'drawing', 'entities', 'keys', 'level', 'sound'], function(audio, drawing, entities, keys, level, sound) {

    var raf = window.requestAnimationFrame || window.webkitRequestAnimationFrame;
    var started;
    var lastDate = Date.now();

    var paused = false;

    // P for pause.
    keys.down.on(80, function() {
        paused = !paused;
        sound.play('select');
        if (paused) {
            drawing.drawPaused();
        }
    });

    // M for mute.
    keys.down.on(77, audio.toggleMute);

    function loop() {
        var now = Date.now();
        var delta = now - lastDate;

        if (!paused) {
            level.getCurrent().tick(delta, level.next, level);
            drawing.draw();
        }

        raf(loop);
        lastDate = now;
    }

    return {
        start: function() {
            if (started) return;
            raf(loop);
            started = true;
        }
    };
});

File: script/timing.dart
import 'dart:html';

import 'drawing.dart' as drawing;
import 'keys.dart' as keys;
import 'levels.dart' as levels;
import 'sound.dart' as sound;


var started = false;
var paused = false;

var previous = 0;

void _loop(num ts) {
    var now = new DateTime.now().millisecondsSinceEpoch;
    var delta = now - previous;

    if (!paused) {
        levels.getCurrent().tick(delta, levels.next);
        drawing.draw();
    }

    window.animationFrame.then(_loop);
    previous = now;
}

void start() {
    if (started) {
        return;
    }
    previous = new DateTime.now().millisecondsSinceEpoch;
    window.animationFrame.then(_loop);
    started = true;

    keys.up.on(80, (e) {
        paused = !paused;
        sound.play('select');

        if (paused) {
            drawing.drawPaused();
        }
    });
}
